internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go

```diff
@@ -140,10 +147,76 @@ func (h *serverSidePatchHelper) Patch(ctx context.Context) error {
 	return h.client.Patch(ctx, h.modified, client.Apply, options...)
 }
 
+// dryRunSSAPatch uses server side apply dry run to determine if the operation is going to change the actual object.
+func dryRunSSAPatch(ctx context.Context, c client.Client, originalUnstructured, modifiedUnstructured *unstructured.Unstructured, helperOptions *HelperOptions) (bool, bool, error) {
+	// For dry run we use the same options as for the intent but with adding metadata.managedFields
+	// to ensure that changes to ownership are detected.
+	dryRunHelperOptions := &HelperOptions{
+		allowedPaths: append(helperOptions.allowedPaths, []string{"metadata", "managedFields"}),
+		ignorePaths:  helperOptions.ignorePaths,
+	}
+
+	// Add TopologyDryRunAnnotation to notify validation webhooks to skip immutability checks.
+	dryRunUnstructured := modifiedUnstructured.DeepCopy()
+	if err := unstructured.SetNestedField(dryRunUnstructured.Object, "", "metadata", "annotations", TopologyDryRunAnnotation); err != nil {
+		return false, false, errors.Wrap(err, "failed to add topology dry-run annotation")
+	}
+
+	// Do a server-side apply dry-run request to get the updated object.
+	err := c.Patch(ctx, dryRunUnstructured, client.Apply, client.DryRunAll, client.FieldOwner(TopologyManagerName), client.ForceOwnership)
+	if err != nil {
+		// This catches errors like metadata.uid changes.
+		return false, false, errors.Wrap(err, "failed to request dry-run server side apply")
+	}
+
+	// Cleanup the dryRunUnstructured object to remove the added TopologyDryRunAnnotation
+	// and remove the affected managedFields  for `manager=capi-topology` which would
+	// otherwise show the additional field ownership for the annotation we added and
+	// the changed managedField timestamp.
+	if err := cleanupTopologyDryRunAnnotation(dryRunUnstructured); err != nil {
+		return false, false, errors.Wrap(err, "failed to filter topology dry-run annotation on dryRunUnstructured")
+	}
+	// Also run the function for the originalUnstructured to remove the managedField
+	// timestamp for `manager=capi-topology`.
+	if err := cleanupTopologyDryRunAnnotation(originalUnstructured); err != nil {
+		return false, false, errors.Wrap(err, "failed to filter topology dry-run annotation on originalUnstructured")
+	}
+
+	// Drop the other fields which are not part of our intent.
+	filterObject(dryRunUnstructured, dryRunHelperOptions)
+	filterObject(originalUnstructured, dryRunHelperOptions)
```

Comment:
Q: Is there a difference between doing this here vs one level above? 
(dryRunUnstructured is a copy of modifiedUnstructured so we already ran filterObject on it, and if we can keep the filterObject of originalUnstructured one level above it feels a bit easier to reason about what's going on)

Of course if the change in allowedPaths is relevant here we can't do that

Reply:
Yes, so no we can't move it up.

For dryRunUnstructured: it gets filled via the client.Patch request.

For originalUnstructured: inside this function the helper preserves the managedFields which are important for detecting ownership changes.